c10/cuda/impl/CUDAGuardImpl.h

```diff
@@ -205,11 +205,35 @@ struct CUDAGuardImpl final : public c10::impl::DeviceGuardImplInterface {
     cuda_stream.synchronize();
   }
 
+  void synchronizeEvent(void* event) const override {
+    if (!event)
+      return;
+    cudaEvent_t cuda_event = static_cast<cudaEvent_t>(event);
+    const c10::impl::PyInterpreter* interp = c10::impl::GPUTrace::get_trace();
+    if (C10_UNLIKELY(interp)) {
+      (*interp)->trace_gpu_event_synchronization(
+          c10::kCUDA, reinterpret_cast<uintptr_t>(cuda_event));
+    }
+    C10_CUDA_CHECK(cudaEventSynchronize(cuda_event));
+  }
+
   void recordDataPtrOnStream(const c10::DataPtr& data_ptr, const Stream& stream)
       const override {
     CUDAStream cuda_stream{stream};
     CUDACachingAllocator::recordStream(data_ptr, cuda_stream);
   }
+
+  double elapsedTime(void* event1, void* event2) const override {
+    TORCH_CHECK(
+        event1 && event2,
+        "Both events must be recorded before calculating elapsed time.");
+    cudaEvent_t cuda_event1 = static_cast<cudaEvent_t>(event1);
+    cudaEvent_t cuda_event2 = static_cast<cudaEvent_t>(event2);
+    float time_ms = 0;
```

Comment:
Could you add the device guard like https://github.com/pytorch/pytorch/blob/1ecea513b648502e4a0765256d165db5d3b42f52/aten/src/ATen/cuda/CUDAEvent.h#L157 here? Looks important.

Reply:
Updated. We also provide support on HIP backend.

Reply:
Why did you add get/set calls here and not `DeviceGuard guard(c10::kCUDA);` ?

Reply:
In my opinion, `CUDAGuardImpl` is an implementation of `DeviceGuard`. `DeviceGuard` is high-level than `CUDAGuardImpl`. So it is unreasonable to use high-level code(`DeviceGuard`) in low-level code(`CUDAGuardImpl`).
If you prefer to use `DeviceGuard`, I can prepare a PR to refine it.